File: src/ae_attribute_inspector/typed_value_parser.rb
module AE


module AttributeInspector


require(File.join(PATH, 'typed_value_parser.rb'))


# Attributes can have values of different types (and nested). This module helps
#validating, parsing and encoding values unambiguously for transmission between
# SketchUp and the WebDialog.
module TypedValueParser


unless defined?(self::COMMA)
  self::Integer = (defined?(::Integer)) ? ::Integer : ::Fixnum
  comma     = /,\s*/
  #float     = /[+-]?\d+\.(?:\d+(?:e[+-]\d+)?)?/
  float     = /[-+]?\d+(?:\.\d+(?:e[+-]\d+)?)?/
  #integer    = /[-]?\d+(?=[^\.])/
  integer    = /[-]?\d+/
  posinteger = /\d+/
  length    = /#{float}(?:mm|cm|m|\"|\')/ # TODO: support length in Vector3d and Point3d
  COMMA     = /^,?\s*/
  NIL       = /^nil/
  BOOLEAN   = /^true|false/
  TIME      = /^(\d{4})-(\d{2})-(\d{2})T([012]\d):([012345]\d):([012345]\d(?:.\d{3}\d{3}?)?)Z/
  #LENGTH    = /^[-]?\d+\.(?:\d+(?:e[+-]\d+)?)?(?:mm|cm|m|\"|\')/
  LENGTH    = /^#{float}(?:mm|cm|m|\"|\')/
  #FLOAT     = /^[-]?\d+\.(?:\d+(?:e[+-]\d+)?)?/
  FLOAT     = /^#{float}/
  #INTEGER    = /^[-]?\d+(?=[^\.])/
  INTEGER    = /^#{integer}/
  STRING    = /^"((?:[^"]|\\")*)"/
  COLOR     = /^Color\((#{posinteger})#{comma}(#{posinteger})#{comma}(#{posinteger})(?:#{comma}(#{posinteger}))?\)/
  POINT3D   = /^Point3d\((#{float})#{comma}(#{float})#{comma}(#{float})\)/
  #VECTOR3D  = /^Vector3d\((#{float})#{comma}(#{float})#{comma}(#{float})\)/
  VECTOR3D  = /^Vector3d\((?:(#{float})#{comma}){2}(#{float})\)/
  FLOAT_3TUPLE  = /^\s*?[\(\[]?(#{float})#{comma}(#{float})#{comma}(#{float})[\)\]]?\s*?$/
  COLOR_34TUPLE = /^\s*?[\(\[]?(#{posinteger})#{comma}(#{posinteger})#{comma}(#{posinteger})(?:#{comma}(#{posinteger}))?[\)\]]?\s*?$/
  #TIME     = /^Time\((#{float})\)/
  SQUARE_BRACKET_START = /^\[/
  SQUARE_BRACKET_END   = /^\]?/
end


def self.determine_type_string(value)
  if value.is_a?(String) && value[/^[\{\[]/]
    begin
      JSON.parse(value)
      return 'String'
    rescue JSON::ParserError
      return value.class.to_s
    end
  elsif value.is_a?(TrueClass) || value.is_a?(FalseClass)
    return 'Boolean'
  elsif value.is_a?(self::Integer)
    return 'Integer'
  else
    return value.class.to_s
  end
end


# Parses a string of this plugin's internal representation of typed values into Ruby objects.
# @param [String] string
# @param [Class,String] type
# @return [String,Boolean,NilClass,Float,Integer,Length,Sketchup::Color,Geom::Point3d,Geom::Vector3d,Time,Array]
# @raises [SyntaxError]
def self.parse(string, type=nil)
  klass = get_class(type) if type.is_a?(String) rescue nil # otherwise type will be determined by parser_implementation
  result = case
  # Some type-specific switches:
  # Strings without escaping or quotes; 
  #    when nested in an array, escaping and quotes are required.
  # Integers, Floats:
  #   when nested in an array, floats must always have a decimal point to 
  #   distinguish them from integers.
  # Colors, points, vectors:
  #   can be written as comma-separated numbers; 
  #   when nested in an array, they must always be wrapped in 
  #     Color(128,128,128,255), Point3d(1,0,0), Vector3d(1,0,0) etc.
  when klass == String then string
  when klass == JSON
    begin
      JSON.parse(string)
    rescue JSON::ParserError => error
      raise(SyntaxError, error)
    end
  when klass == self::Integer
    # Match against a regular expression, if not match then raise syntax error.
    string[INTEGER] || raiseSyntaxError(string, type)
    $&.to_i
  when klass == Float
    string[FLOAT] || raiseSyntaxError(string, type)
    $&.sub(',', '.').to_f
  when klass == Sketchup::Color
    string[COLOR] || string[COLOR_34TUPLE] || raise(SyntaxError)
    if $4.nil?
      Sketchup::Color.new($1.to_i, $2.to_i, $3.to_i)
    else
      Sketchup::Color.new($1.to_i, $2.to_i, $3.to_i, $4.to_i)
    end
  when klass == Geom::Point3d
    string[POINT3D] || string[FLOAT_3TUPLE] || raiseSyntaxError(string, type)
    Geom::Point3d.new($1.to_f, $2.to_f, $3.to_f)
  when klass == Geom::Vector3d
    string[VECTOR3D] || string[FLOAT_3TUPLE] || raiseSyntaxError(string, type)
    Geom::Vector3d.new($1.to_f, $2.to_f, $3.to_f)
  when klass == Length then string.to_l
  else # Boolean, Time, nested types like Array, Hash etc.
    parse_implementation(string, type)
  end
  # After parsing, the remaining string should be empty.
  raise(SyntaxError, "String `#{string}` does not match type #{type}") unless ($'.nil? || $'.empty?)
  return result
end


# Converts typed attribute values into a string of this plugin's internal representation of typed values.
# @param [String,Boolean,NilClass,Float,Integer,Length,Sketchup::Color,Geom::Point3d,Geom::Vector3d,Time,Array] object
# @return [String]
def self.stringify(object)
  return (object.is_a?(String)) ? object : stringify_implementation(object)
end


class << self


  private


  def get_class(class_name_string)
    class_name_string.split('::').reduce(Kernel){ |m, s| m.const_get(s) }
  end


  def raiseSyntaxError(string, type)
    raise SyntaxError.new("String `#{string}` does not match type #{type}")
  end


  # Parses a string of this plugin's internal representation of typed values into Ruby objects.
  # @param [String] string
  # @return [String,Boolean,NilClass,Float,Integer,Length,Sketchup::Color,Geom::Point3d,Geom::Vector3d,Time,Array]
  # @private
  # @raises [SyntaxError]
  def parse_implementation(string, type=nil)
    case string
    when BOOLEAN
      raiseSyntaxError(string, type) if type && type != 'Boolean' && type != 'TrueClass' && type != 'FalseClass'
      return $& == 'true'
    when NIL
      raiseSyntaxError(string, type) if type && type != 'NilClass'
      return nil
    when TIME
      raiseSyntaxError(string, type) if type && type != 'Time'
      usec = ($7.nil?) ? 0 : ($7.length == 3) ? 1000*$7.to_i : ($7.length == 6) ? $7.to_i : 0
      return Time.utc($1.to_i, $2.to_i, $3.to_i,  $4.to_i, $5.to_i, $6.to_i, usec)
      #return Time.at($1.to_f)
    when LENGTH
      raiseSyntaxError(string, type) if type && type != 'Length'
      return $&.to_l
    when FLOAT
      raiseSyntaxError(string, type) if type && type != 'Float'
      return $&.to_f
    when INTEGER
      raiseSyntaxError(string, type) if type && type != 'Integer'
      return $&.to_i
    when STRING
      raiseSyntaxError(string, type) if type && type != 'String'
      return $1.gsub(/\\"/, '"').gsub(/\\\\/, '\\')
    when COLOR
      raiseSyntaxError(string, type) if type && type != 'Sketchup::Color'
      return ($4.nil?) ?
        Sketchup::Color.new($1.to_i, $2.to_i, $3.to_i) : 
        Sketchup::Color.new($1.to_i, $2.to_i, $3.to_i, $4.to_i)
    when POINT3D
      raiseSyntaxError(string, type) if type && type != 'Geom::Point3d'
      return Geom::Point3d.new($1.to_f, $2.to_f, $3.to_f)
    when VECTOR3D
      raiseSyntaxError(string, type) if type && type != 'Geom::Vector3d'
      return Geom::Vector3d.new($1.to_f, $2.to_f, $3.to_f)
    when SQUARE_BRACKET_START
      raiseSyntaxError(string, type) if type && type != 'Array'
      array = []
      until $'[SQUARE_BRACKET_END]
        $'[COMMA]
        array << parse_implementation($')
      end
      return array
    else # nil
      raise(SyntaxError, "TypedValueParser: invalid string `#{string}`")
    end
  end


  # Converts typed attribute values into a string of this plugin's internal representation of typed values.
  # @param [String,Boolean,NilClass,Float,Integer,Length,Sketchup::Color,Geom::Point3d,Geom::Vector3d,Time,Array] object
  # @return [String]
  # @private
  def stringify_implementation(object)
    return case object
      when Time
        (RUBY_VERSION.to_f > 2) ? object.strftime("%Y-%m-%dT%H:%M:%S.%LZ") : object.strftime("%Y-%m-%dT%H:%M:%SZ")
        #"Time(#{value.to_f})"
      when String
        "\"#{object.gsub(/\\/, "\\\\").gsub(/\\"/, '\\"')}\""
      when Sketchup::Color
        "Color(#{object.to_a.map{ |c| c.to_i }.join(", ")})"
      when Geom::Point3d
        "Point3d(#{object.to_a.map{ |c| c.to_f }.join(", ")})"
      when Geom::Vector3d
        "Vector3d(#{object.to_a.map{ |c| c.to_f }.join(", ")})"
      when Array
        "[#{object.map{ |a| stringify_implementation(a) }.join(", ")}]"
      # Boolean, Nil, Length, Float, Integer
      else object.to_s
      end
  end


end # class << self


end # module TypedValueParser


end # module AttributeInspector


end # module AE

File: dialogs/typedvalueparser.js
const COMMA = /^\s*,\s*/
const NIL = /^nil|^null/i
const BOOLEAN = /^true|^false/i
const TIME = /^(\d{4})-(\d{2})-(\d{2})T([012]\d):([012345]\d):([012345]\d(?:.(\d{3}\d{3}?))?)Z/
const LENGTH = /^[-]?\d+(?:\.\d+)?\s*?(mm|cm|m|"|')/
const FLOAT = /^[-+]?\d+[.,](?:\d+(?:e[+-]\d+)?)?/
const INTEGER = /^[-]?\d+/
// const STRING = /^\"((?:[^\"]|\\\")*)\"/
const STRING = /^"((?:[^"]|\\")*)"|'((?:[^']|\\')*)'/
// const COLOR = /^Color\((\d+),\s?(\d+),\s?(\d+)(?:,\s?(\d+))\)/
const COLOR = /^(?:Color|rgb|rgba)\((\d+),\s*?(\d+),\s*?(\d+)(?:,\s*?(\d+))?\)|^#((?:[0-9a-f]{2}){3,4})/
const FLOAT_3TUPLE = /^\s*?[([]?[-+]?\d+(?:\.\d+(?:e[+-]\d+))?,\s*?[-+]?\d+(?:\.\d+(?:e[+-]\d+))?,\s*?[-+]?\d+(?:\.\d+(?:e[+-]\d+))?[)\]]?\s*?$/
const COLOR_34TUPLE = /^\s*?[([]?(\d+),\s*?(\d+),\s*?(\d+)(?:,\s*?(\d+))?[)\]]?\s*?$/
const POINT3D = /^(?:Geom::)?Point3d\(([-]?\d+(?:\.\d+)?),\s?([-]?\d+(?:\.\d+)?),\s?([-]?\d+(?:\.\d+)?)\)/
const VECTOR3D = /^(?:Geom::)?Vector3d\(([-]?\d+(?:\.\d+)?),\s?([-]?\d+(?:\.\d+)?),\s?([-]?\d+(?:\.\d+)?)\)/
const SQUARE_BRACKET_START = /^\[/
const SQUARE_BRACKET_END = /^]/
const NONCOMMA = /^[^,]+/

/**
 * @module TypedValueParser
 * Attributes can have values of different types (and nested). This module helps
 * validating, parsing and encoding values unambiguously for transmission between
 * SketchUp and the WebDialog.
 */
const TypedValueParser = {
  TYPES: [
    'String',
    'JSON',
    'Integer',
    'Float',
    'Length',
    'Array',
    'Boolean',
    'NilClass',
    'Geom::Point3d',
    'Geom::Vector3d',
    'Sketchup::Color',
    'Time',
  ],

  getDisplayName(type) {
    if (/^(Geom|Sketchup)::/.test(type)) {
      return type.split('::')[1]
    } else {
      return type
    }
  },

  /**
   * Validate a value against the given type. Since a type is specified, it
   * tolerates more loose syntax in the input string (ie. Vectors in array syntax).
   * @param {string} string
   * @param {string} type
   * @returns {boolean}
   */
  validate(string, type) {
    string = strip(string)
    switch (type) {
      case 'Boolean':
        return BOOLEAN.test(string) && RegExp['$\''] == ''
      case 'NilClass':
        return NIL.test(string) && RegExp['$\''] == ''
      case 'Time':
        return (
          !isNaN(new Date(string).getTime()) ||
          (TIME.test(string) && RegExp['$\''] == '')
        )
      case 'Length':
        return LENGTH.test(string) && RegExp['$\''] == ''
      case 'Float':
        return FLOAT.test(string) && RegExp['$\''] == ''
      case 'Integer':
        return INTEGER.test(string) && RegExp['$\''] == ''
      case 'String':
        return true // STRING.test(string) && RegExp["$'"] == '';
      case 'JSON':
        try {
          JSON.parse(string)
          return true
        } catch (e) {
          // SyntaxError
          if (e instanceof SyntaxError) {
            return false
          } else {
            throw e
          }
        }
      case 'Sketchup::Color':
        return (
          (COLOR.test(string) || COLOR_34TUPLE.test(string)) &&
          RegExp['$\''] == ''
        )
      case 'Geom::Point3d':
        return (
          (POINT3D.test(string) || FLOAT_3TUPLE.test(string)) &&
          RegExp['$\''] == ''
        )
      case 'Geom::Vector3d':
        return (
          (VECTOR3D.test(string) || FLOAT_3TUPLE.test(string)) &&
          RegExp['$\''] == ''
        )
      case 'Array':
        if (SQUARE_BRACKET_START.test(string)) {
          //while (RegExp["$'"] != '' && !SQUARE_BRACKET_END.test(RegExp["$'"])) {
          while (
            !SQUARE_BRACKET_END.test(RegExp['$\'']) &&
            !(RegExp['$\''] == '')
          ) {
            COMMA.test(RegExp['$\''])
            if (!validateStrict(RegExp['$\''])) {
              return false
            }
          }
          return RegExp['$\''] == ''
        } else {
          return false
        }
      default:
        return false
    }
  },

  /**
   * Parses and transforms a string into a more a consistent format.
   * @param {string} string
   * @param {string} type
   * @returns {string}
   */
  canonicalizeString(string, type) {
    if (type == 'String') {
      return '"' + string.replace(/\\/, '\\\\').replace(/"/, '\\"') + '"'
    }
    string = strip(string)
    switch (type) {
      case 'Sketchup::Color':
        if (COLOR.test(string) || COLOR_34TUPLE.test(string)) {
          if (RegExp.$5 != '') {
            // A match for a hexadecimal code has been found, convert it into rgba.
            let result = []
            for (let i = 0; i < RegExp.$5.length; i += 2) {
              result.push(parseInt(RegExp.$5.slice(i, i + 2), 16))
            }
            return 'Color(' + result.join(',') + ')'
          } else if (RegExp.$3 != '') {
            let result = []
            for (let i = 1; i <= 3; i++) {
              result.push(parseInt(RegExp['$' + i]))
            }
            if (RegExp.$4 != '') {
              result.push(parseInt(RegExp.$4, 16))
            }
            return 'Color(' + result.join(',') + ')'
          } else {
            throw new Error(
              'Function canonicalizeString attempted to exit with null when parsing: ' +
                string
            )
          }
        }
        break
      case 'Geom::Point3d':
        if (POINT3D.test(string) || FLOAT_3TUPLE.test(string)) {
          return (
            'Point3d(' +
            parseFloat(RegExp.$1) +
            ',' +
            parseFloat(RegExp.$2) +
            ',' +
            parseFloat(RegExp.$3) +
            ')'
          )
        }
        break
      case 'Geom::Vector3d':
        if (VECTOR3D.test(string) || FLOAT_3TUPLE.test(string)) {
          return (
            'Vector3d(' +
            parseFloat(RegExp.$1) +
            ',' +
            parseFloat(RegExp.$2) +
            ',' +
            parseFloat(RegExp.$3) +
            ')'
          )
        }
        break
      default:
        return canonicalizeStringStrict(string)
    }
  },

  /**
   * Helper method to convert a JavaScript Date object into a ISO8601 date string.
   * @param {Date} date
   * @returns {string}
   */
  dateToISO8601(date) {
    // TODO: use date.toJSON();
    return (
      ljust(date.getUTCFullYear(), 4) +
      '-' +
      ljust(date.getUTCMonth() + 1, 2) +
      '-' +
      ljust(date.getUTCDate(), 2) +
      'T' +
      ljust(date.getUTCHours(), 2) +
      ':' +
      ljust(date.getUTCMinutes(), 2) +
      ':' +
      ljust(date.getUTCSeconds(), 2) +
      'Z'
    )
  },
}

/**
 * Validate a value.
 * @param {string} string
 * @private
 * @returns {boolean}
 */
function validateStrict(string) {
  if (
    BOOLEAN.test(string) ||
    NIL.test(string) ||
    LENGTH.test(string) ||
    FLOAT.test(string) ||
    INTEGER.test(string) ||
    STRING.test(string) ||
    COLOR.test(string) ||
    POINT3D.test(string) ||
    VECTOR3D.test(string)
  ) {
    return true
  } else if (!isNaN(new Date(string).getTime()) || TIME.test(string)) {
    return true
  } else if (SQUARE_BRACKET_START.test(string)) {
    //while (RegExp["$'"] != '' && !SQUARE_BRACKET_END.test(RegExp["$'"])) {
    while (!SQUARE_BRACKET_END.test(RegExp['$\'']) && !RegExp['$\''] == '') {
      COMMA.test(RegExp['$\''])
      if (!validateStrict(RegExp['$\''])) {
        return false
      }
    }
    return true
  } else {
    return false
  }
}

/**
 * Transform a string into a consistent form.
 * @param {string} string
 * @private
 * @returns {string}
 */
function canonicalizeStringStrict(string) {
  if (BOOLEAN.test(string)) {
    return /true/i.test(RegExp['$&']) ? 'true' : 'false'
  } else if (NIL.test(string)) {
    return 'nil'
  } else if (LENGTH.test(string)) {
    return RegExp.$1 + RegExp.$2
  } else if (TIME.test(string)) {
    return RegExp['$&']
  } else if (FLOAT.test(string)) {
    let float = parseFloat(RegExp['$&'].replace(/,/, '.'))
    return String(float).replace(/^[^.]+$/, '$&.')
  } else if (INTEGER.test(string)) {
    return RegExp['$&']
  } else if (STRING.test(string)) {
    return RegExp['$&']
  } else if (COLOR.test(string)) {
    if (RegExp.$5 != '') {
      // Convert hex code into rgb 255 color
      let result = []
      for (let i = 0; i < RegExp.$5.length; i += 2) {
        result.push(parseInt(RegExp.$5.slice(i, i + 2), 16))
      }
      return 'Color(' + result.join(',') + ')'
    } else if (RegExp.$3 != '') {
      let result = []
      for (let i = 1; i <= 3; i++) {
        result.push(parseInt(RegExp['$' + i], 16))
      }
      if (RegExp.$4 != '') {
        result.push(parseInt(RegExp.$4, 16))
      }
      return 'Color(' + result.join(',') + ')'
    }
    throw new Error(
      'Function canonicalizeString attempted to exit with null when parsing: ' +
        string
    )
  } else if (POINT3D.test(string)) {
    return (
      'Point3d(' +
      parseFloat(RegExp.$1) +
      ',' +
      parseFloat(RegExp.$2) +
      ',' +
      parseFloat(RegExp.$3) +
      ')'
    )
  } else if (VECTOR3D.test(string)) {
    return (
      'Vector3d(' +
      parseFloat(RegExp.$1) +
      ',' +
      parseFloat(RegExp.$2) +
      ',' +
      parseFloat(RegExp.$3) +
      ')'
    )
  } else if (NONCOMMA.test(string)) {
    let d = new Date(RegExp['$&'])
    if (!isNaN(d.getTime())) {
      return self.dateToISO8601(d)
    } else if (SQUARE_BRACKET_START.test(string)) {
      let array = []
      //while (RegExp["$'"] != '' && !SQUARE_BRACKET_END.test(RegExp["$'"])) {
      while (!SQUARE_BRACKET_END.test(RegExp['$\'']) && !RegExp['$\''] == '') {
        COMMA.test(RegExp['$\''])
        let arrayElement = canonicalizeStringStrict(RegExp['$\''])
        if (typeof arrayElement == 'string') {
          array.push(arrayElement)
        }
      }
      return '[' + array.join(',') + ']'
    }
    throw new Error(
      'Function canonicalizeString attempted to exit with null when parsing: ' +
        string
    )
  }
}

export default TypedValueParser

/**
 * Helper method to prepend a number with zeros.
 * @param {Object} o - object or number
 * @oaram {Number} n - The desired total length of the resulting string.
 * @returns {string}
 */
function ljust(o, n) {
  let result = String(o)
  for (let i = result.length; i < n; i++) {
    result = '0' + result
  }
  return result
}

/**
 * Helper method to remove whitespace from beginning and end.
 * @param {string} string
 * @returns {string}
 */
function strip(string) {
  return string.replace(/^\s+/, '').replace(/\s+$/, '')
}
